pymc/tests/test_shape_handling.py

```diff
@@ -304,7 +324,7 @@ def test_can_resize_data_defined_size(self):
         with pm.Model() as pmodel:
             x = pm.MutableData("x", [[1, 2, 3, 4]], dims=("first", "second"))
             y = pm.Normal("y", mu=0, dims=("first", "second"))
-            z = pm.Normal("z", mu=y, observed=np.ones((1, 4)))
+            z = pm.Normal("z", mu=y, observed=np.ones((1, 4)), size=y.shape)
```

Comment:
This behavior was neat, but could cause bugs, because observed may broadcast the shape implied by the parameters

Reply:
Observed variable don't need the shape argument, unless you want to tell PyMC something that it would otherwise not be able to know: i.e., it's shape is the same as another variable or one of it's inputs.

Otherwise you would incorrectly miss these cases:

```python
def test_broadcast_by_observed(self):
        with pm.Model() as m:
            y = pm.Normal("y")
            x = pm.Normal("x", mu=np.zeros((1,)), observed=np.zeros((3,)))
            assert x.eval().shape == (3,)
```
You can't just ignore the shape of observed. Notice that  all this is just a shortcut so the user doesn't have to specify shape when we have observed. But the correct thing is to say shape=observed.shape... unless the user gives us shape directly.

Reply:
I understand the example,  but in multiple of these changes in the tests I don't see why it's necessary.

Is this the behavioral change because you are resizing the RV to the observed shape even if it already matches?
And this unlinks it from the symbolic shape that is implied by the parameters?

I'm not sure if that's a great idea..  This is a non-obvious need to specify shape in order **to turn off** the smart behavior of broadcasting to observed shape..

It would be more consistent to **not** automatically broadcast to the observed shape, but require to specify shape/size in these cases.
This would also mean that the shape is inherited from parameters just like with unobserved variables.

Reply:
> It would be more consistent to not automatically broadcast to the observed shape, but require to specify shape/size in these cases.
This would also mean that the shape is inherited from parameters just like with unobserved variables.

Observed is by definition the shape of the variable, wheres implied dimensions is not. I understand it's not as neat as what we had but the problem is that in Aesara we don't know if parameters are going to have to be broadcasted or not many times.

Reply:
Taken together with the other breaks I think we should do a major version bump then.
This started as a fix of a rare edge case,  but now it is a breaking change that effects potentially many lines of user code.

Reply:
Please edit the PR description to mention this new need for explicit shape/size!

Reply:
> Please edit the PR description to mention this new need for explicit shape/size!

Added this to major/breaking changes:
- When shape is not provided, dims or observed are used to define the shape of a variable. If you want a variable that has dims or is observed to automatically resize when its inputs change, you must specify that explicitly via the shape argument. For example, `pm.Normal("likelihood", mu=mu, sigma=sigma, observed=data, dims="data", shape=mu.shape)`